wandb/integration/ultralytics/callback.py

```diff
@@ -0,0 +1,458 @@
+                "Probabilities",
+                "Speed",
+            ]
+            validation_columns = ["Data-Index", "Batch-Index"] + classification_columns
+            validation_columns.insert(3, "Ground-Truth-Category")
+            train_columns = ["Epoch"] + validation_columns
+            self.train_validation_table = wandb.Table(
+                columns=["Model-Name"] + train_columns
+            )
+            self.validation_table = wandb.Table(
+                columns=["Model-Name"] + validation_columns
+            )
+            self.prediction_table = wandb.Table(
+                columns=["Model-Name"] + classification_columns
+            )
+        elif self.task == "pose":
+            validation_columns = [
+                "Data-Index",
+                "Batch-Index",
+                "Image-Ground-Truth",
+                "Image-Prediction",
+                "Num-Instances",
+                "Mean-Confidence",
+                "Speed",
+            ]
+            train_columns = ["Epoch"] + validation_columns
+            self.train_validation_table = wandb.Table(
+                columns=["Model-Name"] + train_columns
+            )
+            self.validation_table = wandb.Table(
+                columns=["Model-Name"] + validation_columns
+            )
+            self.prediction_table = wandb.Table(
+                columns=[
+                    "Model-Name",
+                    "Image-Prediction",
+                    "Num-Instances",
+                    "Mean-Confidence",
+                    "Speed",
+                ]
+            )
+
+    def _make_predictor(self, model: YOLO):
+        overrides = copy.deepcopy(model.overrides)
+        overrides["conf"] = 0.1
+        self.predictor = self.task_map[self.task]["predictor"](
+            overrides=overrides, _callbacks=None
+        )
+
+    def _save_model(self, trainer: TRAINER_TYPE):
+        model_checkpoint_artifact = wandb.Artifact(
+            f"run_{wandb.run.id}_model", "model", metadata=vars(trainer.args)
+        )
+        checkpoint_dict = {
+            "epoch": trainer.epoch,
+            "best_fitness": trainer.best_fitness,
+            "model": copy.deepcopy(de_parallel(self.model)).half(),
+            "ema": copy.deepcopy(trainer.ema.ema).half(),
+            "updates": trainer.ema.updates,
+            "optimizer": trainer.optimizer.state_dict(),
+            "train_args": vars(trainer.args),
+            "date": datetime.now().isoformat(),
+            "version": __version__,
+        }
+        checkpoint_path = trainer.wdir / f"epoch{trainer.epoch}.pt"
+        torch.save(checkpoint_dict, checkpoint_path, pickle_module=pickle)
+        model_checkpoint_artifact.add_file(checkpoint_path)
+        wandb.log_artifact(
+            model_checkpoint_artifact, aliases=[f"epoch_{trainer.epoch}"]
+        )
+
+    def on_train_start(self, trainer: TRAINER_TYPE):
+        with telemetry.context(run=wandb.run) as tel:
```

Comment:
is it fair to assume that we have a run at this point?

Reply:
Yes, because the run is initialized inside the callback event `on_pretrain_routine_start` by [`ultralytics`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/callbacks/wb.py#L26).